syz-fuzzer/proc.go

```diff
@@ -89,6 +90,16 @@ func (proc *Proc) loop() {
 			p := proc.fuzzer.target.Generate(proc.rnd, prog.RecommendedCalls, ct)
 			log.Logf(1, "#%v: generated", proc.pid)
 			proc.execute(proc.execOpts, p, ProgNormal, StatGenerate)
+		} else if proc.fuzzer.mabSSEnabled {
+			// Use MAB engine to pick a prog.
+			// There is no need to use snapshot here as MAB engine has its own lock.
+			// Also, the size of proc.fuzzer.corpus can only increase.
+			pidx, pr := fuzzerSnapshot.mabSS.Choose(proc.rnd)
```

Comment:
I can see there is a concurrency problem here: fuzzerSnapshot and mabSS.Choose is protected by different mutexes and it's possible for mabSS's status to change (since it's a pointer) after fuzzerSnapshot is taken. We can either
1) Let fuzzerSnapshot has a deep copy of mabSS (including its MAB engine) or
2) Let mabSS maintain its own copy of seed corpus array.

Any suggestions?

Reply:
I have modified the seed selection engine to maintain its own corpus. This way we don't have to create snapshots in two different places and risk race conditions. It's also helpful for the synchronization process since the fuzzer needs to access the seed program for synchronization.

Reply:
What exactly is the problem?
I see that we need the chosen program index pidx to be not larger than the corpus size. This can be ensured by taking snapshot after selecting a program I think.
The corpus passed to p.Mutate does not need to be consistent with anything.